### ISSUE TITLE
lesson 3 articles and research to read

### ISSUE THREAD

**Issue by datalass1** [datalass1/fastai#11]
### learning rates:
https://towardsdatascience.com/estimating-optimal-learning-rate-for-a-deep-neural-network-ce32f2556ce0
https://miguel-data-sc.github.io/2017-11-05-first/
### pytorch:
https://towardsdatascience.com/a-practitioners-guide-to-pytorch-1d0f6a238040
### advanced: 
http://teleported.in/posts/decoding-resnet-architecture/

**Comment by datalass1** [comment]
### Estimating an Optimal Learning Rate For a Deep Neural Network 
**Pavel Surmenok**

**Comment by datalass1** [comment]
### Visualizing Learning rate vs Batch size, Neural Nets basics using Fast.ai tools
![BatchSize](https://miguel-data-sc.github.io/img/Captura6.PNG)

**Comment by datalass1** [comment]
### A practitioner's guide to PyTorch
**Radek Osmulski**

Tensor — (like) a numpy.ndarray but can live on the GPU.

Variable — allows a tensor to be part of a computation by wrapping itself around it. If created with requires_grad = True, will have gradients calculated during the backwards phase.

Things to keep in mind (or else risk going crazy)
1. Datatypes matter!
2. If it can overflow or underflow, it will. Visit [AI With The Best Oct 2017](https://www.youtube.com/watch?v=XlYD8jn1ayE&feature=youtu.be&list=PLoWh1paHYVRfygApBdss1HCt-TFZRXs0k&t=1607) for knowledge of numerical stability. 
3. Gradients accumulate by default! By default, gradients accumulate. You run a computation once, you run it backwards — everything is fine. But, for the second run, the gradients get added to the gradients from the first operation! 

Visit [PyTorch offical tutorials](https://pytorch.org/tutorials/)

**Comment by datalass1** [comment]
### Decoding the ResNet architecture
**Anand Saha**

The takeaway is that you should not be using smaller networks because you are afraid of overfitting. Instead, you should use as big of a neural network as your computational budget allows, and use other regularization techniques to control overfitting -  cs231n

It is well known that increasing the depth leads to exploding or vanishing gradients problem if weights are not properly initialized. However, that can be countered by techniques like batch normalization.

**Make it deep, but remain shallow**
Given a shallower network - how can we take it, add extra layers and make it deeper - without losing accuracy or increasing error? It’s tricky to do but one insight is that if the extra layers added to the deeper network are identity mappings, they become equivalent to the shallower network. And hence, they should produce no higher training error than it’s shallower counterpart. This is called a solution by construction.

![shallowvsdeepnetwork](http://teleported.in/post_imgs/12-shallow-deep.png "A Shallow network (left) and a deeper network (right) constructed by taking the layers of the shallow network and adding identity layers between them. Img Credit: Kaiming He.")
Paper to read about [Deep Residual Learning for Image Recognition](https://arxiv.org/pdf/1512.03385.pdf).

**Understanding residual**
A residual is the error in a result.
Let’s say, you are asked to predict the age of a person, just by looking at her. If her actual age is 20, and you predict 18, you are off by 2. 2 is our residual here. If you had predicted 21, you would have been off by -1, our residual in this case. In essence, residual is what you should have added to your prediction to match the actual.
